README.md
# Mobify Bellows

A mobile-first accordion UI module for progressive disclosure on the web.

[![NPM](https://nodei.co/npm/bellows-ui.png?downloads=true)](https://nodei.co/npm/bellows-ui/)
[![Dependency Status](https://www.versioneye.com/user/projects/53ff9b7deab62a5003000009/badge.svg?style=flat)](https://www.versioneye.com/user/projects/53ff9b7deab62a5003000009)
[![Build Status](https://circleci.com/gh/mobify/bellows/tree/master.svg?style=shield&circle-token=8c096b513a1f04ef1b977178c4f3358c2a42a132)](https://circleci.com/gh/mobify/bellows)

![Bellows in action](https://raw.githubusercontent.com/mobify/bellows/master/examples/assets/i/bellows.gif "Bellows in action")

## Demo

You can find a simple demo on [the Documentation page](http://mobify.github.io/bellows). More demos can be found inside the `examples` folder in the repo. Run `grunt examples` to see them in Chrome (mobile device emulation is recommended).

## Requirements

* [jQuery](http://jquery.com/)
* [Velocity.js](http://julian.com/research/velocity/)
* [Mobify's Plugin Factory](https://www.npmjs.com/package/mobify-plugin/)

### Zepto Support

Bellows supports Zepto up until v5.1.2 but is not actively developed for it. While we don't actively support Zepto for Bellows, we welcome any and all issues and PRs to help us make it work.


## Installation

Bellows can be installed using NPM:

```
npm install bellows-ui
```

## Usage with Require.js

To use with require.js, after installing through NPM you merely have to reference bellows in your require config file (**Note**: If your project already has these external dependencies, and the versions are compatible, it's recommended that you use the one in your project to reduce duplication):

```config.js

{
    'paths': {
        '$': 'node_modules/bellows-ui/node_modules/jquery/dist/jquery.min',
        'plugin': 'node_modules/bellows-ui/node_modules/mobify-plugin/dist/plugin.min',
        'velocity': 'node_modules/bellows-ui/node_modules/velocity-animate/velocity'
        'bellows': 'node_modules/bellows-ui/dist/bellows.min',
    }
}

```

And then require bellows in as needed:

```
define(
    ['$', 'bellows'],
    function($) {
        $('.bellows').bellows();
    }
);
```


## Usage

At a bare minimum, your markup structure should follow the above structure. You should have at least one `bellows__item`. Content within `bellows__header` and `bellows__content` can be whatever you want. You may also style either of those however you need. Our default theme will give you some standard styling for those sections but, if you want to theme Bellows yourself, we recommend not including the theme file and starting from scratch.

```html
<!-- Include the CSS -->
<link rel="stylesheet" href="bellows.min.css">

<!-- Optionally include the Theme file -->
<link rel="stylesheet" href="bellows-style.min.css">

<!-- Include the markup -->
<div class="bellows">
    <!-- The Accordion Items -->
    <div class="bellows__item">
        <div class="bellows__header">
            <!-- Item Header - Content can be whatever you want -->
        </div>
        <div class="bellows__content">
            <!-- Item Content - Content can be whatever you want -->
        </div>
    </div>
    <div class="bellows__item">
        <div class="bellows__header">
            <h3>Header</h3>
        </div>
        <div class="bellows__content">
            <p>Content</p>
        </div>
    </div>
    <!-- Use the disabled class if you want items to not respond to open/closed events -->
    <div class="bellows__item bellows--is-disabled">
        <div class="bellows__header">
            <h3>Header</h3>
        </div>
        <div class="bellows__content">
            <p>Content</p>
        </div>
    </div>
</div>

<!-- Include dependencies -->
<script src="jqueryin.js"></script>
<script src="node_modules/velocity-animate/velocity.min.js"></script>
<script src="node_modules/mobify-plugin/dist/plugin.min.js"></script>

<!-- Include bellows.js -->
<script src="bellows.min.js"></script>

<!-- Construct Bellows -->
<script>$('.bellows').bellows()</script>
```

## Initializing the plugin

### bellows()

Initializes the bellows.

```js
$('.bellows').bellows();
```

### bellows(options)

Initialize with options.

```js
$('.bellows').bellows({
    singleItemOpen: false,
    duration: 200,
    easing: 'swing',
    open: function(e, ui) {},
    opened: function(e, ui) {},
    close: function(e, ui) {},
    closed: function(e, ui) {}
});
```

#### Options

##### singleItemOpen

default: `false`

When set to `true` will force only one item open at a time.

```js
$('.bellows').bellows({
    singleItemOpen: true
});
```

##### duration

default: `200`

Sets the duration for the animation.

```js
$('.bellows').bellows({
    duration: 600
});
```

##### easing

default: `swing`

Sets the easing for the animation. Bellows takes all of the same easing properties that [Velocity.js](http://julian.com/research/velocity) accepts.

> * [jQuery UI's easings](http://easings.net/) and CSS3's easings ("ease", "ease-in", "ease-out", and "ease-in-out"), which are pre-packaged into Velocity. A bonus "spring" easing (sampled in the CSS Support pane) is also included.
* CSS3's bezier curves: Pass in a four-item array of bezier points. (Refer to [Cubic-Bezier.com](http://cubic-bezier.com/) for crafing custom bezier curves.)
* Spring physics: Pass a two-item array in the form of [ tension, friction ]. A higher tension (default: 600) increases total speed and bounciness. A lower friction (default: 20) increases ending vibration speed.
* Step easing: Pass a one-item array in the form of [ steps ]. The animation will jump toward its end values using the specified number of steps.

For more information, check out [Velocity's docs on easing](http://julian.com/research/velocity/#easing).

```js
$('.bellows').bellows({
    easing: 'ease-in-out'
});
```

##### open

default: `function(e, ui) {}`

Triggered every time the selected bellows item is starting to open.

**Parameters**

| Parameter name | Description |
|----------------|-------------|
| **e** | An Event object passed to the callback |
| **ui** | An object containing any associated data for use inside the callback |

```js
$('.bellows').bellows({
    open: function(e, ui) {
        // ui.item contains the item opening
    }
});
```

##### opened

default: `function(e, ui) {}`

Triggered every time the selected bellows item has finished opening.

**Parameters**

| Parameter name | Description |
|----------------|-------------|
| **e** | An Event object passed to the callback |
| **ui** | An object containing any associated data for use inside the callback |

```js
$('.bellows').bellows({
    opened: function(e, ui) {
        // ui.item contains the item that opened
    }
});
```

##### close

default: `function(e, ui) {}`

Triggered every time an bellows item is starting to close.

| Parameter name | Description |
|----------------|-------------|
| **e** | An Event object passed to the callback |
| **ui** | An object containing any associated data for use inside the callback |

```js
$('.bellows').bellows({
    close: function(e, ui) {
        // ui.item contains the item closing
    }
});
```

##### closed

default: `function(e, ui) {}`

Triggered every time an bellows item is finished closing.

| Parameter name | Description |
|----------------|-------------|
| **e** | An Event object passed to the callback |
| **ui** | An object containing any associated data for use inside the callback |

```js
$('.bellows').bellows({
    closed: function(e, ui) {
        // ui.item contains the item that closed
    }
});
```

## Methods

### open

Open the selected bellows item by element reference

```js
$bellows.bellows('open', $('.bellows__item'));
```

or by index

```js
$bellows.bellows('open', 1);
```

### openAll

Opens all the bellows items

```js
$bellows.bellows('openAll');
```

### close

Close the selected bellows item by element reference

```js
$bellows.bellows('close', $('.bellows__item'));
```

or by index

```js
$bellows.bellows('close', 1);
```

### closeAll

Closes all the bellows items

```js
$bellows.bellows('closeAll');
```

### toggle

Toggle the selected bellows item by element reference

```js
$bellows.bellows('toggle', $('.bellows__item'));
```

or by index

```js
$bellows.bellows('toggle', 1);
```

### toggleAll

Toggles all the bellows items

```js
$bellows.bellows('closeAll');
```

### add

Adds new items to bellows, and correctly wraps their content elements. Optional third parameter allows you to replace all existing elements with the ones specified.

```js
$bellows.bellows('add', items);
```

or replacing existing elements

```js
$bellows.bellows('add', items, true);
```

## Known Issues and Workarounds

When the `singleItemOpen` configuration option is selected and the bellows items contain content that is larger than the overall viewport, closing an item can leave the user disoriented. This is due to the way `singleItemOpen` works; it closes all open items, then animates the selected one open. This can cause the positioning of the active item in the viewport to shift either up or down, resulting it it not being in the same position as when it was clicked.

To get around this, you can scroll to the item once the animation finishes. To do so, you just need to add the following to the `opened` event:

```
$('.bellows').bellows({
    singleItemOpen: true,
    opened: function(e, ui) {
        Velocity.animate(ui.item, 'scroll');
    }
});
```

This will scroll the viewport to the opened bellows item, restoring its position in the viewport.

## Browser Compatibility

| Browser           | Version | Support                    |
|-------------------|---------|----------------------------|
| Mobile Safari     | 4.0.x   | Degraded. No transitions.  |
| Mobile Safari     | 5.0+    | Supported.                 |
| Android Browser   | 4.0+    | Supported.                 |
| Android Browser   | 2.3.x   | Degraded. No transitions.  |
| Chrome (Android)  | 1.0+    | Supported.                 |


## Building a distribution

### Requirements
* [Node.js v4.x LTS + NPM v2.x](https://nodejs.org/en/download/) (Mobify recommends [NVM](https://github.com/creationix/nvm) for installing Node + NPM)
* [Grunt](http://gruntjs.com/)
    * Install with `npm install -g grunt-cli`

### Steps
1. `npm install`
1. `grunt build`

The `dist` directory will be populated with minified versions of the css and javascript files for distribution and use with whatever build system you might use. The `src` directory has our raw unminified Sass and Javascript files if you prefer to work with those.

## License

_MIT License. Bellows is Copyright © 2016 Mobify. It is free software and may be redistributed under the terms specified in the LICENSE file._


CHANGELOG
6.0.1
  - Fix [#62](https://github.com/mobify/bellows/issues/62): Use published Plugin dependency
  - Fix [#63](https://github.com/mobify/bellows/issues/63): Short circuit open if other items are opening
6.0.0
  - Publish to NPM
  - Remove Zepto support in favor of jQuery
  - Migrate all dependencies to NPM, remove Bower
  - Migrate JS linting to eslint
  - Migrate SCSS compilation to libsass
  - Lint SCSS with sass-lint
  - Update circleCI config with modern testing steps
  - Add some GitHub templates for contributing, issues, and PRs
5.1.2
  - Ensuring disabled items don't close on close
5.0.1
  - Reworked test infrastructure to use more robust setup and teardown using iframes
5.0.0
  - Upgrading to latest version of plugin factory
  - Adding `destroy` method
  - Removing `event` option in favour of internalizing a namespaced event
4.1.2
  - Moved gremlins.js dependency to devDependencies
4.1.1
  - Updated version of Velocity to fix queuing issues
4.1.0
  - Added ability to disable bellows items
4.0.1
  - Updated bellows to use new mobify-velocity fork
4.0.0
  - Fixed breaking issue with plugin factory
3.0.0
  - Refactored to use new plugin factory
2.0.3
  - Implemented event delegation instead of direct bellows item binding
  - Allow inclusion of navigable anchors inside header items
  - Added linter to ensure correct code standards
  - Added new `add` method to dynamically add/replace bellows items
  - Refactored CSS class constants into object
  - Added some comments for clarity
2.0.2
  - Velocity had a 1.0 release. This brings Bellows into line with the latest version of Velocity.
2.0.1
  - Fix for issue #34 - aria-hidden attributes behavior is reversed
2.0.0
  - Complete rewrite of Bellows. Everything is new. Everything is better.


src/js/bellows.js
(function(factory) {
    if (typeof define === 'function' && define.amd) {
        define([
            '$',
            'velocity',
            'plugin'
        ], factory);
    } else {
        var framework = window.jQuery;
        factory(framework, framework.Velocity, window.Plugin);
    }
})(function($, Velocity, Plugin) {
    var cssClasses = {
        ITEM: 'bellows__item',
        HEADER: 'bellows__header',
        OPENED: 'bellows--is-open',
        OPENING: 'bellows--is-opening',
        CLOSING: 'bellows--is-closing',
        DISABLED: 'bellows--is-disabled'
    };

    var selectors = {
        ITEM_HEADER: '> .bellows__item > .bellows__header',
        ITEM_CONTENT_WRAPPER: '> .bellows__content-wrapper',
        ITEM_CONTENT: '> .bellows__item > .bellows__content'
    };

    var events = {
        CLICK: 'click.bellows'
    };

    var Bellows = function(element, options) {
        Bellows.__super__.call(this, element, options, Bellows.DEFAULTS);
    };

    Bellows.VERSION = '0';

    Bellows.DEFAULTS = {
        singleItemOpen: false,
        duration: 200,
        easing: 'swing',
        open: $.noop,
        opened: $.noop,
        close: $.noop,
        closed: $.noop
    };

    Plugin.create('bellows', Bellows, {

        _init: function(element) {
            this.$bellows = $(element);

            this._wrapContent(this.$bellows);

            this._bindEvents();
        },

        destroy: function() {
            this.$bellows
                .removeData(this.name)
                .off(events.CLICK);
        },

        _bindEvents: function() {
            var plugin = this;

            /**
             * Ghetto Event Delegation™
             TODO: Re-evalute this for new default of jQuery
             Zepto doesn't support descendant selectors in event delegation,
             so we compare against the closest bellows to ensure we are invoking
             the event from a direct child, not a bellows child from a nested bellows.
             */
            this.$bellows
                .on(events.CLICK, function(e) {
                    var $target = $(e.target);
                    var $closestBellows = $target.closest('.bellows');

                    // We don't want to continue if we're clicking on an anchor
                    if ($target.is('a') || !!$target.parents('a').length) {
                        return;
                    }

                    // We need to verify not only that we're inside the direct bellows of the item, but also if the item is a header/child of a header
                    if ($closestBellows[0] === plugin.$bellows[0] && ($target.hasClass(cssClasses.HEADER) || !!$target.closest('.bellows__header').length)) {
                        e.preventDefault();

                        plugin.toggle($target.closest('.bellows__item'));
                    }
                });
        },

        /*
         Gets an element's height using Velocity's built-in property cache.
         Used for getting heights before animations, for animating into an
         element's space.
         */
        _getHeight: function($element) {
            return parseFloat(Velocity.CSS.getPropertyValue($element[0], 'height'));
        },

        /*
         Sets the height of bellows so we animate into
         the space rather than re-flowing the entire document
         */
        _setHeight: function(height) {
            this.$bellows.css('height', height || '');
        },

        /*
         Allow items to be found using an index
         */
        _item: function(item) {
            if (typeof item === 'number') {
                item = this.$bellows.find('.' + cssClasses.ITEM).eq(item);
            }

            return item;
        },

        _isOpen: function($item) {
            return $item.hasClass(cssClasses.OPENED);
        },

        _isDisabled: function($item) {
            return $item.hasClass(cssClasses.DISABLED);
        },

        _wrapContent: function($items) {
            $items
                .find(selectors.ITEM_CONTENT)
                // wrap content section of each item to facilitate padding
                .wrap('<div class="bellows__content-wrapper" />')
                // add aria-hidden attribute to all hidden content wrappers
                .parents('.bellows__item:not(.bellows--is-open)')
                .find(selectors.ITEM_CONTENT_WRAPPER)
                .attr('aria-hidden', true);
        },

        /*
        Checks if other bellows items are currently opening
        */
        _othersAreOpening: function() {
            return this.$bellows.find('.' + cssClasses.OPENING).length > 0;
        },

        toggle: function($item) {
            $item = this._item($item);

            this[$item.hasClass(cssClasses.OPENED) ? 'close' : 'open']($item);
        },

        open: function($item) {
            $item = this._item($item);

            if (this._isOpen($item) || this._isDisabled($item)) {
                return;
            }

            var plugin = this;
            var $contentWrapper = $item.find(selectors.ITEM_CONTENT_WRAPPER);

            if (this.options.singleItemOpen) {
                this.closeAll();

                // Fix #63
                if (this._othersAreOpening()) {
                    return;
                }
            }

            this._trigger('open', {item: $item});

            Velocity
                .animate($contentWrapper, 'slideDown', {
                    begin: function() {
                        plugin._setHeight(plugin._getHeight(plugin.$bellows) + plugin._getHeight($contentWrapper));
                        $item.addClass(cssClasses.OPENING);
                    },
                    duration: this.options.duration,
                    easing: this.options.easing,
                    complete: function() {
                        $item
                            .removeClass(cssClasses.OPENING)
                            .addClass(cssClasses.OPENED)
                            .find(selectors.ITEM_CONTENT_WRAPPER)
                            .removeAttr('aria-hidden');

                        plugin._setHeight();

                        plugin._trigger('opened', {item: $item});
                    }
                });
        },

        close: function($item) {
            $item = this._item($item);

            if (!this._isOpen($item) || this._isDisabled($item)) {
                return;
            }

            var plugin = this;
            var $contentWrapper = $item.find(selectors.ITEM_CONTENT_WRAPPER);

            this._trigger('close', {item: $item});

            Velocity
                .animate($contentWrapper, 'slideUp', {
                    begin: function() {
                        plugin._setHeight(plugin._getHeight(plugin.$bellows));
                        $item
                            .removeClass(cssClasses.OPENED)
                            .addClass(cssClasses.CLOSING);
                    },
                    duration: this.options.duration,
                    easing: this.options.easing,
                    complete: function() {
                        $item
                            .removeClass(cssClasses.CLOSING)
                            .find(selectors.ITEM_CONTENT_WRAPPER)
                            .attr('aria-hidden', true);

                        plugin._setHeight();

                        plugin._trigger('closed', {item: $item});
                    }
                });
        },

        toggleAll: function() {
            var plugin = this;

            this.$bellows.find('.' + cssClasses.ITEM).each(function() {
                plugin.toggle($(this));
            });
        },

        openAll: function() {
            var plugin = this;

            this.$bellows.find('.' + cssClasses.ITEM + ':not(.' + cssClasses.OPENED + ')').each(function() {
                plugin.open($(this));
            });
        },

        closeAll: function() {
            var plugin = this;

            this.$bellows.find('.' + cssClasses.OPENED).each(function() {
                plugin.close($(this));
            });
        },

        add: function(items, replace) {
            var $container = $('<div />');
            $(items).appendTo($container);

            replace && this.$bellows.empty();

            this._wrapContent($container);
            this.$bellows.append($container.children());
        }
    });

    $('[data-bellows]').bellows();

    return $;
});
